### PR TITLE
Several patches

### PR DESCRIPTION
**Pull request by fsb4000**
>This introduces the concept of the 'sync node', which is the one we
asked for missing blocks. In case the sync node goes away, a new one
will be selected.

>For now, the heuristic is very simple, but it can easily be extended
later to add better policies.

Same as https://github.com/novacoin-project/novacoin/commit/28f9882707d389250e307ebf58dcf981340f1381
Similar to https://github.com/bitcoin/bitcoin/commit/6ed71b5e4faa2ea6d739eefa9c3dcc15db80439c
(I don't delete old sync code because it does conflict with nothing and that's OK "Ask the first connected node for block updates" before this node can be selected for block updates in net.cpp)
(Although I can remove old sync code if necessary.)
Issue description: https://github.com/bitcoin/bitcoin/issues/1234

### CODE REVIEW

**Comment by fsb4000** [comment]
I added "using pthread_create instead CreateThread on Windows". Not sure that it is somehow connected with the problem, but it's the only thing that I did Windows only.(for compatibility with mingw 4.9.2)
Check whether it does break WinBuilder or not...

**Comment by fsb4000** [comment]
Maybe https://github.com/chronokings/huntercoin/issues/6 fixed. Need additional tests...

**Comment by fsb4000** [comment]
OpenSSL Fix = https://github.com/bitcoin/bitcoin/pull/5634 and https://github.com/bitcoin/bitcoin/pull/5640

**Comment by fsb4000** [comment]
Make lsn_reset ("detach databases") optional and off by default = https://github.com/bitcoin/bitcoin/pull/1119

**Comment by domob1812** [comment]
Thanks for the pull request!  In general, please don't submit "several patches" but split them into logical units (for the future).

Also, without having reviewed the code in full, I have at least some suggestions right now:

1) The lsn_reset patch seems fine, but I would prefer to have lsn_reset *on* by default.  Otherwise people may get problems when they try to backup / restore / transfer their data folder and have no idea why.  If they are aware of the possible implications, then they can turn it off for themselves.

2) Please do not include the hardcoded nodes.  I don't think that's a good idea (do we have more info about the nodes, like their capacity, who runs them, how they are secured).  We have IRC discovery for that, and may add DNS seeds in the future.

3) Please do not introduce checkpoints.  It was a deliberate decision not to have them (not by me, but it was made).  Also, you don't even mention that in your forum post.  That's a very good demonstration of why I suggest to have multiple pull requests for multiple things, and not just a generic "several patches".

File: src/huntercoin.cpp [hunk 1]
```diff
@@ -2927,8 +2943,20 @@ unsigned short GetDefaultPort()
...
+the decimal address.
+Example:
+IP address = 67.225.171.185
+First Octet: 	67
+Second Octet: 	225
+Third Octet: 	171
+Fourth Octet: 	185
+(67 * 16777216) + (225 * 65536) + (171 * 256) + (185) = 1138863033
+*/
+unsigned int pnSeed[] = {2990412266, 1138863033 };
```
Comment (domob1812) [inline src/huntercoin.cpp]: Why add the comment and not just #define a macro that *does* the calculation for you?

(Besides, I don't think we should hardcode the nodes.)

File: src/huntercoin.cpp [hunk 2]
```diff
@@ -2945,3 +2973,4 @@ string GetDefaultDataDirSuffix() {
 }
 
 unsigned char GetAddressVersion() { return ((unsigned char)(fTestNet ? 100 : 40)); }
+
```
Comment (domob1812) [inline src/huntercoin.cpp]: Nit pick: Please remove the blank line.

**Comment by fsb4000** [comment]
@domob1812:
>Nit pick: Please remove the blank line.

added to the last commit(Automatic creation huntercoin.conf) I do not know how to edit an arbitrary commit, only the last. Let me know if you know how to edit an arbitrary commit.

@domob1812:
>Why add the comment and not just #define a macro that does the calculation for you?
(Besides, I don't think we should hardcode the nodes.)

Without adding nodes, huntercoin does not start synchronization from scratch. Although it works, if user downloads the database from your site(http://chain.huntercoin.org/)(because your database has addr.dat with node IPs)
Of course I can add a macro. Could you prompt me a name for this macro?

**Comment by domob1812** [comment]
Just add another commit which cleans up everything.  (This is not perfect, but I think it is good enough for now.)

Actually, Huntercoin should discover nodes via IRC (as Bitcoin did back in the day).  Is this broken?  If so, then we should indeed try to fix it.  If not, I don't think we should be adding the hardcoded nodes.

**Comment by fsb4000** [comment]
IRC code is good, but IRC servers often is offline(or some thing)
Some times I get
**IRC connect failed** at debug.log...
**IRC waiting 71 seconds to reconnect**
**IRC connect failed**
**IRC waiting 138 seconds to reconnect**
and so. I think better use IRC + harcode seeds + hardcode dns seeds...

**Comment by fsb4000** [comment]
@domob1812:
>Do we have more info about the nodes, like their capacity, who runs them, how they are secured?

Yes, we have. User can't synchronize. The official advices were:
**BGB:**
>Yes this has been a problem. You can try adding my node
addnode=67.225.171.185

**Snailbrain:**
>you can try  this node:- 178.62.17.234
but have limited to 40 connections max and all used atm..
but if you give me your ip and port is open, i can connect=you - i believe this should work even though i'm maxed.

So I added 67.225.171.185 and 178.62.17.234 in the code.

**Comment by domob1812** [comment]
Ok, I see.  In that case, if IRC is down sometimes (wasn't aware of that), I think adding the nodes makes sense.  Please just use a macro like

#define CONVERT_IP4(a, b, c, d) \
  (((a) << 24) + ((b) << 16) + ((c) << 8) + (d))

instead of the manual computation.

**Comment by fsb4000** [comment]
@domob1812:
> Please do not introduce checkpoints. It was a deliberate decision not to have them (not by me, but it was made). Also, you don't even mention that in your forum post. That's a very good demonstration of why I suggest to have multiple pull requests for multiple things, and not just a generic "several patches".

Done, https://github.com/fsb4000/huntercoin/commit/6b7eed1d458ea9d17fbbcc2823cd1f71cb777834

**Comment by fsb4000** [comment]
@domob1812 :
>The lsn_reset patch seems fine, but I would prefer to have lsn_reset on by default. Otherwise people may get problems when they try to backup / restore / transfer their data folder and have no idea why. If they are aware of the possible implications, then they can turn it off for themselves.

done, https://github.com/fsb4000/huntercoin/commit/1ea795dd5055c331793686a6f155c6a5c2ebb2b5

**Snailbrain**, please,in the next release which will include the pull request, provide details about command line command **-detachdb=0** which really speeds up the client shutdown
(-detachdb=0 for huntercoind and uncheck checkbox "Detach databases at shutdown" for Qt version)
detach

**Comment by fsb4000** [comment]
"Automatic creation huntercoin.conf" is similar to https://github.com/jl777/btcd/commit/be2f7323faba13d7ad729c1f988ee2effd235401

**Pull request merged**